subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java

```diff
@@ -76,17 +77,18 @@ public File populate(File versionDir, VersionRef ref, VersionControlSpec spec) {
             // The HEAD reference in a Git Repository is a logical choice if the user is looking
             // for the 'latest.integration' version of a dependency.
             if (gitRef.getVersion().equals("HEAD")) {
-                versions.add(GitVersionRef.from("latest.integration", gitRef.getCanonicalId()));
+                versions.add(GitVersionRef.from(VersionRef.LATEST_INTEGRATION, gitRef.getCanonicalId()));
             }
         }
         return versions;
     }
 
-    private static void cloneRepo(File workingDir, GitVersionControlSpec gitSpec) {
+    private static void cloneRepo(File workingDir, GitVersionControlSpec gitSpec, VersionRef ref) {
         CloneCommand clone = Git.cloneRepository().setURI(gitSpec.getUrl().toString()).setDirectory(workingDir);
         Git git = null;
         try {
             git = clone.call();
+            git.reset().setMode(ResetCommand.ResetType.HARD).setRef(ref.getCanonicalId()).call();
```

Comment:
I don't understand why we're resetting the repository?

Reply:
If we clone, we get the latest version. In our case, we have a commit that went in between the resolution of the dependencies. This means that when we will clone, we will get the latest commit and not the commit we actually want to resolve. Resetting ensure we get the right commit. I'm not a Git expert so this may not be optimal.

Reply:
My problem is that I think populate() should effectively be a no-op. We already cloned or updated the working directory when the first external dependency was resolved. So, this second one for the same `GitVersionControlSpec` and the same `VersionRef` in the same `versionDir` shouldn't cause any work to be done.

Rather than cloning or updating, maybe we should just say. Oh, we've already populated the repository at that version. I guess we don't need to do anything.

Reply:
Discussed this with @lacasseio. He's right. We do need to reset the repo to the commit we are targeting.